### ISSUE TITLE
inference(ensemble): add ensemble-of-MAP/VI runner via vmap over PRNG keys

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#70]
## Problem / Request

Add a generic **ensemble-of-MAP / ensemble-of-VI runner** to `pyrox.inference` that fits `E` independent point estimates (MAP) or mean-field surrogates (VI) in a single `vmap` over PRNG keys. The runner is backend-agnostic (consumes a `numpyro` model or a raw log-joint) and returns params stacked on a leading ensemble axis so downstream code can treat the ensemble as a mixture.

This is the cheapest useful uncertainty estimator we can offer: no HMC, no NUTS, no trace-replay — just parallel point-mass posteriors. It is also the inference substrate that the sklearn-style `EstimatorBase` facade (follow-up issue) and the Bayesian Neural Field layer family (follow-up issue) both build on.

## User Story

As a modeler who wants ballpark predictive uncertainty on a `numpyro`-or-equinox model without paying the cost of MCMC, I want to fit an ensemble of `E` MAP (or VI) estimates with one call, get back a PyTree whose leaves have a leading `(E,)` axis, and then vmap my predictive over that axis to get an ensemble of predictions.

## Motivation

Three things are missing from the current `pyrox` inference surface:

1. **Cheap uncertainty.** `pyrox.gp` ships conjugate + sparse paths, but for neural models users only have MLE-style `optax` loops. The ensemble-of-MAP recipe is the minimum-viable-Bayes substitute — it underpins bayesnf, deep-ensembles (Lakshminarayanan 2017), and most "Bayesian-ish" NN benchmarks.
2. **A common entrypoint for MAP / MLE / VI.** Users currently write three slightly different SGD loops for `MLE = MAP with uniform prior`, `MAP = full log-joint`, and `VI = ELBO with mean-field Normal`. A single runner with `prior_weight` and `kind={"map","vi"}` collapses these to one call site.
3. **A building block.** The sklearn-style `Estimator.fit` facade and `BNFEstimator*` need *some* inference engine underneath. Either we invent it inside the facade or we expose a reusable one — the latter is strictly cheaper to test and document.

## Proposed API

```python
# src/pyrox/inference/_ensemble.py

from typing import Callable
import equinox as eqx
import jax
import jax.numpy as jnp
from jaxtyping import Array, Float, PRNGKeyArray, PyTree


def ensemble_map(
    log_joint: Callable[[PyTree, Array, Array], Float[Array, ""]],
    init_fn: Callable[[PRNGKeyArray], PyTree],
    *,
    ensemble_size: int,
    num_epochs: int,
    data: tuple[Array, Array],
    seed: PRNGKeyArray,
    batch_size: int | None = None,
    learning_rate: float = 5e-3,
    prior_weight: float = 1.0,          # 0.0 -> MLE
    optimizer: "optax.GradientTransformation | None" = None,
) -> tuple[PyTree, Float[Array, "E T"]]:
    """Fit `ensemble_size` MAP/MLE estimates in parallel by vmap over PRNG keys.

    `prior_weight` scales the prior term in `log_joint`. `0.0` recovers MLE.
    Returns `(params_stacked, loss_history)` where every leaf of `params_stacked`
    carries a leading `(E,)` axis.
    """


def ensemble_vi(
    model_fn: Callable,                  # numpyro model (x, y=None) -> None
    guide_fn: Callable,                  # numpyro guide
    *,
    ensemble_size: int,
    num_epochs: int,
    data: tuple[Array, Array],
    seed: PRNGKeyArray,
    kl_weight: float = 1.0,              # tempering; 1.0 = standard ELBO
    learning_rate: float = 5e-3,
    batch_size: int | None = None,
    num_particles: int = 1,
) -> tuple[PyTree, Float[Array, "E T"]]:
    """Fit `ensemble_size` mean-field `numpyro` surrogates in parallel via SVI.
    Returns `(guide_params_stacked, loss_history)`.
    """
```

### Sampling back from a fitted ensemble

```python
def ensemble_predict(
    params_stacked: PyTree,
    predict_fn: Callable[[PyTree, Array], Array],   # scalar-in, scalar-out
    x_new: Array,
) -> Float[Array, "E *batch"]:
    """`jax.vmap(predict_fn, in_axes=(0, None))(params_stacked, x_new)`."""
```

This is a three-liner but worth shipping so the user never has to remember which axis is the ensemble axis.

### MAP with `prior_weight`

Given a log-joint $\log p(\theta, y \mid x) = \log p(y \mid x, \theta) + \log p(\theta)$, the runner minimizes the tempered negative log-posterior per ensemble member $e$:

$$\mathcal{L}_e(\theta_e) = -\frac{N}{\lvert B \rvert}\sum_{i \in B} \log p(y_i \mid x_i, \theta_e) \;-\; w_{\text{prior}} \, \log p(\theta_e)$$

where $w_{\text{prior}} \in \mathbb{R}_{\ge 0}$ is the prior weight. Setting $w_{\text{prior}} = 0$ recovers MLE; setting $w_{\text{prior}} = 1$ recovers MAP. The $N / \lvert B \rvert$ rescaling makes the expected mini-batch loss an unbiased estimate of the full-data loss, following the standard SGD-for-MAP convention.

### Ensemble as mixture posterior

Each of the $E$ members converges to a (possibly different) local mode $\hat\theta_e$. Treating the ensemble as a uniform mixture gives

$$q(\theta) \;=\; \frac{1}{E}\sum_{e=1}^{E} \delta\!\bigl(\theta - \hat\theta_e\bigr),$$

so the predictive becomes $p(y^* \mid x^*) \approx \tfrac{1}{E}\sum_e p(y^* \mid x^*, \hat\theta_e)$. This is the Lakshminarayanan et al. (2017) deep-ensembles predictive, and is also what bayesnf uses as its de-facto posterior approximation.

### VI with tempered ELBO

Each surrogate $q_{\phi_e}(\theta)$ maximizes

$$\text{ELBO}(\phi_e) \;=\; \mathbb{E}_{\theta \sim q_{\phi_e}}\bigl[\log p(y \mid x, \theta)\bigr] \;-\; \beta \, \mathrm{KL}\!\bigl(q_{\phi_e} \,\|\, p\bigr),$$

where $\beta$ is `kl_weight`. $\beta < 1$ (cold posterior) often improves predictive calibration on NN-scale models; $\beta = 1$ is the standard ELBO.

## Reference Code (Google's bayesnf)

The original implementation in Google's Bayesian-Neural-Field library is a good sanity reference, but we deliberately diverge from it on three axes: (a) we use `vmap` over keys rather than `pmap(vmap(...))` (pmap is easy to add later), (b) we accept a `numpyro` model or a raw log-joint instead of a `tfp` `JointDistributionSequential`, (c) we do not bake in data shapes — both `features` and `target` ride through as opaque PyTrees.

Relevant excerpts:

```python
# bayesnf/inference.py — ensemble_map skeleton
def ensemble_map(features, target, neg_energy_fn, prior_d, init_fn,
                 ensemble_size, learning_rate, num_epochs, seed,
                 batch_size=None, prior_weight=1.0):
    def _target_log_prob_fn(params, x_batch, y_batch):
        if prior_weight == 0.0:
            return -(neg_energy_fn(params, x_batch, y_batch)
                     * (target.shape[0] / batch_size))
        return -(neg_energy_fn(params, x_batch, y_batch) * (target.shape[0] / batch_size)
                 + prior_d.log_prob(params) * prior_weight)

    init_params = jax.vmap(jax.vmap(init_fn))(
        jax.random.split(init_seed, (num_devices, ensemble_size)))

    @jax.pmap
    @jax.vmap
    def _run(init_params, seed):
        opt = optax.adam(learning_rate)
        ...
        (params, _, _), losses = jax.lax.scan(_one_epoch, ...)
        return params, losses

    return _run(init_params, jax.random.split(opt_seed, (num_devices, ensemble_size)))
```

([source: `bayesnf/inference.py:510-623`](https://github.com/google/bayesnf/blob/main/src/bayesnf/inference.py#L510-L623))

The VI variant is the analogous construction layered over `tfp.vi.fit_surrogate_posterior_stateless` ([`inference.py:626-759`](https://github.com/google/bayesnf/blob/main/src/bayesnf/inference.py#L626-L759)). In our port we use `numpyro.infer.SVI` + `numpyro.infer.Trace_ELBO` directly, and let the user supply `guide_fn` (usually `AutoNormal(model_fn)`).

### Our port — sketch

```python
# src/pyrox/inference/_ensemble.py — sketch
def ensemble_map(log_joint, init_fn, *, ensemble_size, num_epochs,
                 data, seed, batch_size=None, learning_rate=5e-3,
                 prior_weight=1.0, optimizer=None):
    import optax
    x, y = data
    N = y.shape[0]
    B = batch_size if batch_size is not None else N
    opt = optimizer if optimizer is not None else optax.adam(learning_rate)

    def loss_fn(params, xb, yb):
        ll, lp = log_joint(params, xb, yb)           # returns (loglik, logprior)
        return -((N / B) * ll + prior_weight * lp)

    init_keys = jax.random.split(seed, ensemble_size)

    @jax.vmap
    def run_one(init_key):
        params0 = init_fn(init_key)
        state0  = opt.init(params0)

        def epoch(carry, _):
            params, state, key = carry
            key, sub = jax.random.split(key)
            if B < N:
                idx = jax.random.permutation(sub, N)[:B]
                xb, yb = x[idx], y[idx]
            else:
                xb, yb = x, y
            loss, grads = jax.value_and_grad(loss_fn)(params, xb, yb)
            updates, state = opt.update(grads, state)
            params = optax.apply_updates(params, updates)
            return (params, state, key), loss

        (params, _, _), losses = jax.lax.scan(epoch, (params0, state0, init_key),
                                              None, length=num_epochs)
        return params, losses

    return run_one(init_keys)
```

For `ensemble_vi`, the inner `run_one` delegates to `numpyro.infer.SVI` with its stateful `SVIRunResult`, but otherwise follows the same `vmap`-over-keys skeleton.

## References

1. Lakshminarayanan, B., Pritzel, A., & Blundell, C. (2017). *Simple and Scalable Predictive Uncertainty Estimation using Deep Ensembles.* NeurIPS 2017. [arXiv:1612.01474](https://arxiv.org/abs/1612.01474)
2. Saad, F. A., Burnim, J., Carroll, C., Patton, B., Köster, U., Saurous, R. A., & Hoffman, M. D. (2024). *Scalable spatiotemporal prediction with Bayesian neural fields.* Nature Communications 15, 7942. [DOI:10.1038/s41467-024-51477-5](https://doi.org/10.1038/s41467-024-51477-5). Code: [google/bayesnf](https://github.com/google/bayesnf) (see [`inference.py`](https://github.com/google/bayesnf/blob/main/src/bayesnf/inference.py)).
3. Wenzel, F., et al. (2020). *How Good is the Bayes Posterior in Deep Neural Networks Really?* ICML 2020 — motivates the `kl_weight` ($\beta$) tempering knob. [arXiv:2002.02405](https://arxiv.org/abs/2002.02405)
4. Gustafsson, F. K., Danelljan, M., & Schön, T. B. (2020). *Evaluating Scalable Bayesian Deep Learning Methods for Robust Computer Vision.* CVPR-W 2020 — baselines DeepEnsembles vs MFVI on the same scaffold we're building. [arXiv:1906.01620](https://arxiv.org/abs/1906.01620)
5. [numpyro SVI docs](https://num.pyro.ai/en/stable/svi.html) — the underlying VI engine the `ensemble_vi` variant wraps.
6. [optax](https://optax.readthedocs.io/) — gradient transformations; `optax.adam` is the default optimizer.

## Implementation Steps

- [ ] Create `src/pyrox/inference/__init__.py` and `src/pyrox/inference/_ensemble.py`.
- [ ] Implement `ensemble_map` with `vmap` over PRNG keys, `lax.scan` over epochs, mini-batch permutation with `batch_size=None` fallback to full-batch, and the `prior_weight=0.0 → MLE` shortcut.
- [ ] Implement `ensemble_vi` wrapping `numpyro.infer.SVI` + `Trace_ELBO`, exposing `kl_weight` by scaling the likelihood term the same way bayesnf does (so the optimizer still sees a well-conditioned objective).
- [ ] Implement `ensemble_predict` as a thin `vmap(..., in_axes=(0, None))` over a user-supplied scalar-in predictive.
- [ ] Re-export `ensemble_map`, `ensemble_vi`, `ensemble_predict` through `src/pyrox/inference/__init__.py` and `src/pyrox/__init__.py`.
- [ ] Add tests (see Testing).
- [ ] Add an mkdocstrings API stub at `docs/api/inference.md` and list it in `mkdocs.yml`.

## Definition of Done

- `pyrox.inference.ensemble_map` and `pyrox.inference.ensemble_vi` are importable and documented.
- An `E=32` ensemble MAP fit on a toy linear-regression log-joint recovers the analytic ridge-regression posterior mean within `rtol=1e-2` and the ensemble standard deviation shrinks proportionally to $1/\sqrt{N}$ as the training set grows.
- An `ensemble_vi` fit on the same toy problem returns a stacked `guide_params` PyTree whose leaves carry a leading `(E,)` axis, and an `SVI`-loss history of shape `(E, T)` that decreases monotonically on average.
- The BNF estimator follow-up issue is unblocked: `BNFEstimator.fit` can be implemented as a one-line call into `pyrox.inference.ensemble_map`.

## Testing

- **Unit** — `ensemble_map` on a closed-form problem:
  - Log-joint = Gaussian likelihood + Gaussian prior ⇒ analytic ridge solution. Check ensemble-mean vs analytic mean and ensemble-std shrinkage vs $1/\sqrt{N}$.
  - `prior_weight=0.0` ⇒ MLE; compare against `jnp.linalg.lstsq` on the same data.
  - `batch_size` smaller than `N`: check each epoch consumes the correct number of minibatches.
- **Unit** — `ensemble_vi` on the same linear problem:
  - Assert SVI loss decreases on average over the last 10% of epochs.
  - Assert returned `guide_params` leaves have leading `(E,)` axis.
- **Integration** — Validate against an existing `pyrox.gp` model: fit a `GPPrior`/`ConditionedGP` via ensemble MAP with a tight Gamma prior on lengthscale, confirm posterior mean matches `numpyro.infer.MCMC` within tolerance on a synthetic dataset.
- **Determinism** — same seed ⇒ bit-identical `params_stacked`.

## Documentation

- mkdocstrings stub `docs/api/inference.md` listing the three public functions.
- Docstring-level examples showing the `MLE` (`prior_weight=0`), `MAP` (`prior_weight=1`), and `VI` call patterns side-by-side.
- Add a short "Ensemble MAP/VI" section to the top-level docs index cross-linking to the BNF notebook.

## Relationships

- Unlocks: sklearn-style `EstimatorBase` facade (next issue), `BNFEstimator*` family, and future ensemble-style uncertainty estimators in `pyrox.nn`.
- Related prior art: bayesnf's ensemble runner (see References[2]), `numpyro.infer.SVI` (upstream VI engine we wrap), Lakshminarayanan et al. 2017 (deep ensembles baseline).
- Cross-repo: no hard dep on [gaussx#121](https://github.com/jejjohnson/gaussx/issues/121) — but the BNF downstream consumer of *this* issue does need `gaussx.mixture_quantile` for NB/ZINB quantile summaries.
